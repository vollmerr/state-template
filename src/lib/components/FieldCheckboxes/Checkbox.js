import React from 'react';
import T from 'prop-types';

import * as propUtils from '../../utils/propTypes';

// handle changing array value in redux form
export const handleChange = ({ value, values, onChange }) => (event) => {
  const newValues = [...values];
  // first element is blank, remove it
  if (newValues[0] === '') {
    newValues.shift();
  }
  // is checked, add it to array
  if (event.target.checked) {
    newValues.push(value);
  } else {
    // not checked, remove from array
    newValues.splice(newValues.indexOf(value), 1);
  }
  // update in redux form
  return onChange(newValues);
};

// individual checkbox, applies state-template styling
const Checkbox = (props) => {
  const {
    name,
    value,
    option,
    variant,
    disabled,
    onChange,
    'aria-invalid': ariaInvalid,
    'aria-describedby': ariaDescribedby,
  } = props;

  const values = Array.isArray(value) ? value : [value];

  const id = `${name}-${option.value}`;
  const inputProps = {
    id,
    name,
    disabled,
    type: 'checkbox',
    checked: values.includes(option.value),
    onChange: handleChange({ value: option.value, values, onChange }),
    'aria-invalid': ariaInvalid,
    'aria-describedby': ariaDescribedby,
  };

  if (variant) {
    return (
      <label data-test={'field--checkbox'} className={'check'} htmlFor={id}>
        <input className={'hidden-up pos-abs top-0 left-0'} {...inputProps} />
        <span className={`btn btn-md btn-block color-white--checked bg-${variant}--checked rounded-0`}>
          {option.label}
        </span>
      </label>
    );
  }

  return (
    <label data-test={'field--checkbox'} className={'form-check-inline p-l-md m-l-0 m-r-md'} htmlFor={id}>
      <input className={'hidden-up pos-abs'} {...inputProps} />
      <div className={'check-icon-checkbox'} aria-hidden>
        <i className={'ca-gov-icon-check-mark'} />
      </div>
      {option.label}
    </label>
  );
};

Checkbox.propTypes = {
  /** Name of field */
  name: T.string.isRequired,

  /** Value of checkbox */
  value: T.oneOfType([
    T.string,
    T.array,
  ]).isRequired,

  /** Option to select */
  option: propUtils.option.isRequired,

  /** Use style variant */
  variant: T.oneOf([
    '',
    'primary',
  ]),

  /** Disable the input */
  disabled: T.bool,

  /** Label to render for this checkbox */
  label: T.string,

  /** Called when checkbox changes */
  onChange: T.func.isRequired,

  /** Accessible indicator for errors existing */
  'aria-invalid': T.string.isRequired,

  /** Accessible indicator of related information */
  'aria-describedby': T.string,
};

Checkbox.defaultProps = {
  variant: '',
  disabled: false,
  label: null,
  'aria-describedby': null,
};

export default Checkbox;
