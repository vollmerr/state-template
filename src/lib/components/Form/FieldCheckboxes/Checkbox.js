import React from 'react';
import T from 'prop-types';

// handle changing array value in redux form
export const handleChange = ({ value, values, onChange }) => (event) => {
  const newValues = [...values];
  // first element is blank, remove it
  if (newValues[0] === '') {
    newValues.shift();
  }
  // is checked, add it to array
  if (event.target.checked) {
    newValues.push(value);
  } else {
    // not checked, remove from array
    newValues.splice(newValues.indexOf(value), 1);
  }
  // update in redux form
  return onChange(newValues);
};

// individual checkbox, applies state-template styling
const Checkbox = ({
  input, value, label, variant, disabled,
}) => {
  const values = Array.isArray(input.value) ? input.value : [input.value];
  const checked = values.includes(value);
  const onChange = handleChange({ value, values, onChange: input.onChange });

  const inputProps = {
    disabled,
    checked,
    onChange,
    type: 'checkbox',
    name: input.name,
  };

  if (variant) {
    return (
      <label data-test={'field-checkbox'} className={'check'}>
        <input className={'hidden-up pos-abs top-0 left-0'} {...inputProps} />
        <span className={`btn btn-md btn-block color-white--checked bg-${variant}--checked rounded-0`}>{label}</span>
      </label>
    );
  }

  return (
    <label data-test={'field-checkbox'} className={'form-check-inline p-l-md m-l-0 m-r-md'}>
      <input className={'hidden-up pos-abs'} {...inputProps} />
      <div className={'check-icon-checkbox'}>
        <i className={'ca-gov-icon-check-mark'} />
      </div>
      {label}
    </label>
  );
};

Checkbox.propTypes = {
  /** Use style variant */
  variant: T.oneOf([
    '',
    'primary',
  ]),
  /** Input from redux-form's Field, attaches name, value, etc  */
  input: T.object.isRequired,
  /** Value of checkbox */
  value: T.string.isRequired,
  /** Label to display for this checkbox */
  label: T.string.isRequired,
  /** Disable the input */
  disabled: T.bool,
};

Checkbox.defaultProps = {
  variant: '',
  disabled: false,
};

export default Checkbox;
