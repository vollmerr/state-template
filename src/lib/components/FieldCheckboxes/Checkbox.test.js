import React from 'react';
import { shallow } from 'enzyme';

import Checkbox, { handleChange } from './Checkbox';

const props = {
  name: 'test-name',
  value: 'value-1',
  option: {
    label: 'label-1',
    value: 'value-1',
  },
  variant: null,
  disabled: false,
  onChange: jest.fn(),
  'aria-invalid': 'false',
  'aria-describedby': 'aria-desc-id',
};

let wrapper;
describe('Checkbox', () => {
  beforeEach(() => {
    wrapper = shallow(<Checkbox {...props} />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should allow variants', () => {
    wrapper.setProps({ variant: 'primary' });
    const className = wrapper.find('span').prop('className');
    expect(className).toMatch(/bg-primary/);
  });

  it('should be `checked` for matching string values', () => {
    wrapper.setProps({ value: props.value });
    expect(wrapper.find('checked')).toBeTruthy();
  });

  it('should be `checked` for matching values in an array', () => {
    wrapper.setProps({ value: [props.value] });
    expect(wrapper.find('checked')).toBeTruthy();
  });

  let onChange;
  describe('handleChange', () => {
    beforeEach(() => {
      onChange = jest.fn();
    });

    it('should remove blank values from the start of the array', () => {
      const event = { target: { checked: false } };
      handleChange({ value: 'v1', values: ['', 'v 1'], onChange })(event);
      expect(onChange).toBeCalledWith([]);
    });

    it('should add a value to the end of the array if checked', () => {
      const event = { target: { checked: true } };
      handleChange({ value: 'v 1', values: [], onChange })(event);
      expect(onChange).toBeCalledWith(['v 1']);
    });

    it('should remove a value if not checked', () => {
      const event = { target: { checked: false } };
      handleChange({ value: 'v 2', values: ['v 1', 'v 2', 'v 3'], onChange })(event);
      expect(onChange).toBeCalledWith(['v 1', 'v 3']);
    });
  });
});
