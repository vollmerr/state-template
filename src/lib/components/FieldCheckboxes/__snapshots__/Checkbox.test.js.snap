// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render correctly 1`] = `
<label
  className="form-check-inline p-l-md m-l-0 m-r-md"
  data-test="field--checkbox"
  htmlFor="test-name-value-1"
>
  <input
    aria-describedby="aria-desc-id"
    aria-invalid="false"
    checked={true}
    className="hidden-up pos-abs"
    disabled={false}
    id="test-name-value-1"
    name="test-name"
    onChange={[Function]}
    type="checkbox"
  />
  <div
    aria-hidden={true}
    className="check-icon-checkbox"
  >
    <i
      className="ca-gov-icon-check-mark"
    />
  </div>
  label-1
</label>
`;
